gravitee-gateway-flow/src/test/java/io/gravitee/gateway/flow/BestMatchPolicyResolverTest.java

```diff
@@ -0,0 +1,153 @@
+        mockInputRequestPath("/turtle");
+
+        List<Flow> flows = List.of(
+            buildFlow("/book"),
+            buildFlow("/book/:bookId"),
+            buildFlow("/book/:bookId/chapter/:chapterId"),
+            buildFlow("/book/:bookId/chapter")
+        );
+        when(flowResolver.resolve(context)).thenReturn(flows);
+
+        List<Flow> filteredFlows = policyResolver.resolve(context);
+
+        assertEquals(0, filteredFlows.size());
+    }
+
+    @Test
+    public void resolve_should_return_the_flow_with_deeper_matching_path() {
+        mockInputRequestPath("/book/9999/chapter/145");
+
+        List<Flow> flows = mockFlows(
+            buildFlow("/book"),
+            buildFlow("/book/9999/chapter/145/page"),
+            buildFlow("/book/7777/chapter/145"),
+            buildFlow("/book/9999/chapter/147"),
+            buildFlow("/book/9999/chapter/145"),
+            buildFlow("/book/9999/chapter/148"),
+            buildFlow("/book/9999/chapter")
+        );
+
+        List<Flow> filteredFlows = policyResolver.resolve(context);
+
+        assertEquals(1, filteredFlows.size());
+        assertSame(flows.get(4), filteredFlows.get(0));
+    }
+
+    @Test
+    public void resolve_should_return_the_flow_with_deeper_matching_path_with_variables() {
+        mockInputRequestPath("/book/9999/chapter/145");
+
+        List<Flow> flows = mockFlows(
+            buildFlow("/book"),
+            buildFlow("/book/:bookId/chapter/:chapterId/page/:pageId"),
+            buildFlow("/book/:bookId"),
+            buildFlow("/book/:bookId/chapter/:chapterId"),
+            buildFlow("/book/:bookId/chapter")
+        );
+
+        List<Flow> filteredFlows = policyResolver.resolve(context);
+
+        assertEquals(1, filteredFlows.size());
+        assertSame(flows.get(3), filteredFlows.get(0));
+    }
+
+    @Test
+    public void resolve_should_not_care_of_trailing_separator() {
+        mockInputRequestPath("/book/9999/chapter/145/");
+
+        List<Flow> flows = mockFlows(buildFlow("/book/9999/chapter/145"));
+
+        List<Flow> filteredFlows = policyResolver.resolve(context);
+
+        assertEquals(1, filteredFlows.size());
+        assertSame(flows.get(0), filteredFlows.get(0));
+    }
+
+    @Test
+    public void resolve_should_return_multiple_flows_matching_with_same_depth() {
+        mockInputRequestPath("/book/9999/chapter/145");
+
+        List<Flow> flows = mockFlows(
+            buildFlow("/book"),
+            buildFlow("/book/:bookId/chapter/:chapterId"),
+            buildFlow("/book/:bookId"),
+            buildFlow("/book/:bookId/chapter/145"),
+            buildFlow("/book/:bookId/chapter")
+        );
+
+        List<Flow> filteredFlows = policyResolver.resolve(context);
+
+        assertEquals(2, filteredFlows.size());
+        assertSame(flows.get(1), filteredFlows.get(0));
+        assertSame(flows.get(3), filteredFlows.get(1));
+    }
```

Comment:
so.. it really should ? is it what we expect ? :p

Reply:
:question: waiting for architecture team explanations about this point

Reply:
I would say that it should not return the both flows but only the one that is matching best. In this test it should be the one in position 3 in the array.

I'm wondering how it behave with the following 2 additional entries:
`/book/9999/chapter/145`
`/book/:bookId/chapter/145`

Reply:
I added those 2, and resolve() returns 4 flows.

I was expecting BestMatchPolicyResolver to always select only one flow, but that's not it.
Is there any reason why it can return multiple ?
I wonder what is the gateway behavior in that case.

It seems matching algorithm can be improved. In our example, to prioritize plain-matching flows against variables-matching flows

Reply:
As discussed together, we merge those tests as it.

And we continue investigations on :
- is there matching algorithm improvements needed ?
- what happens when more than 1 flow is returned ?
- should it always return one only flow ?